content/blog/eng/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline.md

```diff
@@ -0,0 +1,322 @@
+
+```yaml
+name: # Docker image to be used as service
+alias: # optional hostname for your service in the internal networking
+entrypoint: # optional override for the Docker entrypoint
+command: # optional ovveride for the Docker entrypoint
+```
+
+**But... Isn't there something missing from what we had in our `docker-compose.yaml` file?**
+
+The answer is yes: before [**GitLab** 14.8](https://about.gitlab.com/releases/2022/02/22/gitlab-14-8-released/), `services` couldn't specify a set of custom `environment variables` (as reported in [issue #23671](https://gitlab.com/gitlab-org/gitlab/-/issues/23671) on **GitLab**).
+
+>ℹ️
+>
+>This change is not mentioned in the 14.8 release notes because it's been split over multiple releases, but you can refer to both [**GitLab** 14.5 changelog](https://gitlab.com/gitlab-org/gitlab/-/blob/master/CHANGELOG.md#1450-2021-11-19) and [**GitLab Runner** 14.8 changelog](https://gitlab.com/gitlab-org/gitlab-runner/blob/14-8-stable/CHANGELOG.md#new-features)
+
+So, how are we supposed to add `environment variables` to our **GitLab**'s `services`?
+**GitLab** `jobs` support the `variables` property which allows us to define a set of `environment variables` as a `YAML hash`, and `services` inherit all the variables defined in their parent `job`.
+
+Let's make a quick example of what this means:
+
+```yaml
+job-1:
+    variables:
+        VAR: Job variable # This is meant as a variable to be used in job's script
+        SVC_VAR: Service variable # This is meant as a variable to be used EXCLUSIVELY by the service
+    services:
+        - name: busybox:latest # Use a simple Busybox image so that we can a shell and just echo some variables
+          entrypoint: [ "sh", "-c", "echo $VAR, $SVC_VAR" ]
+    script:
+        - ...
+```
+
+Upon `service` startup the `entrypoint` is executed, leading to the following output:
+
+```
+2022-04-04T12:56:48.722815200Z Job variable, Service variable
+```
+
+>ℹ️
+>
+>`service` output is printed at the beginning of your `job`'s output on **GitLab**. Additionaly, it can be extracted from **Docker** or **Kubernetes** container logs based on where your **GitLab Runner** is installed.
+
+While this seem to work correctly, we have to remind about the warning that I mentioned in the previous chapter: we're using a single **mock server** which is configured using a set of `environment variables`, so each instance of the **mock server** requires exactly the very same `variables`.
+
+Following from the previous example, let's move to something closer to our scenario by tanslating the `docker-compose.yaml` file into a `.gitlab-ci.yml` one:
+
+```yaml
+job-1:
+    services:
+        - name: $MOCK_IMAGE:$COMMIT_ID
+          alias: company-1-mock
+        - name: $MOCK_IMAGE:$COMMIT_ID
+          alias: company-2-mock
+    script:
+        - npm run tests # Generic wrapper to test all the companies
+    variables:
+        COMPANY_NAME: company-1 ⚠️
+        COMPANY_NAME: company-2 ⚠️
```

Comment:
What in the YAML is this stuff? :laughing: Doesn't it trigger any validation failure on GitLab CI's side?

Reply:
I don't know if it's valid or not but it was just a way to mark those two lines as problematic. If you have a better way to make them stand out I'm all ears 😄

Reply:
No no sorry, I wasn't talking about the emojis (which is a great way to highlight, I agree!), I was talking about the duplicated key in the dict!

Reply:
Well, in that case I can tell you that it's actually valid. GitLab will only take the last value in the merged YAML file.

For the sake of convenience, here's the final generated YAML:

```yaml
---
job-1:
  services:
  - name: "$MOCK_IMAGE:$COMMIT_ID"
    alias: company-1-mock
  - name: "$MOCK_IMAGE:$COMMIT_ID"
    alias: company-2-mock
  script:
  - npm run tests
  variables:
    COMPANY_NAME: company-2 ⚠️
```

(as a side note, I just learned that emojis are completely file in that scenario 😄)

Reply:
> (as a side note, I just learned that emojis are completely file in that scenario smile)

Yep, the beauty of UTF-8 :laughing: technically you can even use them in code!